### PR TITLE
Setup test harness with initial test

### LINKED ISSUE #8: Setup test harness and run/validate deploy scripts
- [ ] Set up test harness which forks mainnet, runs the deploy script, and validates deployment

### PR DESCRIPTION
**Pull request by alexkeating**
#### Description

- Add test harness
- Add test for initial contract

Closes #8

### CODE REVIEW

**Comment by alexkeating** [comment]
Current dependencies on/for this PR:
* main
  * **PR #6** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/6" target="_blank">Graphite</a> 
    * **PR #7** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/7" target="_blank">Graphite</a> 
      * **PR #9** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/9" target="_blank">Graphite</a>  👈
        * **PR #11** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/11" target="_blank">Graphite</a> 
          * **PR #12** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/12" target="_blank">Graphite</a> 
            * **PR #13** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/13" target="_blank">Graphite</a> 
              * **PR #14** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/14" target="_blank">Graphite</a> 
                * **PR #15** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/15" target="_blank">Graphite</a> 
                  * **PR #17** <a href="https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/17" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/pooltogether-governor-upgrade/9?utm_source=stack-comment).

**Comment by alexkeating** [comment]
@apbendi Can you please add an github action env var called `MAINNET_RPC_URL`, `DEPLOYER_PRIVATE_KEY`, and `PROPOSER_PRIVATE_KEY`.

File: src/interfaces/IPOOL.sol [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.18;
+
+interface IPOOL {
```
Comment (alexkeating) [inline src/interfaces/IPOOL.sol:4]: *(arbitrary comment)*

Same as gitcoin with a few methods missing

File: test/PooltogetherGovernor.t.sol [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    assertEq(governorBravo.timelock(), TIMELOCK);
+    assertEq(governorBravo.COUNTING_MODE(), "support=bravo&quorum=for,abstain&params=fractional");
+  }
+}
+
+// TODO: Should we make this work better with scopelint. Wrap the
+// inherited methods?
+//
+// Constructor should handle this and we should have use deployed
+// as a flag.
```
Comment (alexkeating) [inline test/PooltogetherGovernor.t.sol]: Should we change the testing naming to fit closer to `scopelint spec`? Is this even possible.

File: test/PooltogetherGovernorTest.sol [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import {PooltogetherGovernor} from "src/PooltogetherGovernor.sol";
+
+abstract contract PooltogetherGovernorTest is Test, DeployInput {
+  using FixedPointMathLib for uint256;
+
+  uint256 constant QUORUM = 2_500_000e18;
+  address constant POOL_TOKEN = 0x0cEC1A9154Ff802e7934Fc916Ed7Ca50bDE6844e;
+  IPOOL poolToken = IPOOL(POOL_TOKEN);
+  address constant TIMELOCK = 0x42cd8312D2BCe04277dD5161832460e95b24262E;
+  address constant PROPOSER = 0xe0e7b7C5aE92Fe94D2ae677D81214D6Ad7A11C27; // lonser.eth
```
Comment (alexkeating) [inline test/PooltogetherGovernorTest.sol]: Taken from Gitcoin these values have been replaced except for `QUORUM`

File: src/interfaces/IPOOL.sol (hunk 1 above)
Comment (apbendi) [reply src/interfaces/IPOOL.sol:4]: Did you verify this with `cast interface` on the actual POOL deployment?

File: test/PooltogetherGovernor.t.sol [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    assertEq(governorBravo.name(), "Pooltogether Governor Bravo");
+    assertEq(address(governorBravo.token()), POOL_TOKEN);
+    // These values were all copied directly from the mainnet alpha governor at:
+    // 0xB3a87172F555ae2a2AB79Be60B336D2F7D0187f0
+    assertEq(INITIAL_VOTING_DELAY, 1);
+    assertEq(INITIAL_VOTING_PERIOD, 28_800);
+    assertEq(INITIAL_PROPOSAL_THRESHOLD, 1000e18);
+    // forgefmt: disable-end
+    assertEq(governorBravo.votingDelay(), INITIAL_VOTING_DELAY);
+    assertEq(governorBravo.votingPeriod(), INITIAL_VOTING_PERIOD);
```
Comment (apbendi) [inline test/PooltogetherGovernor.t.sol:14]: Let's add an assertion that any delays are set below some upper bound threshold of reasonable time (maybe 6 weeks or something like that). One way an error could be made is if someone were to miscalculate while updating these values and set them to something impossibly long. Testing would all work but in real life the Governor could be bricked for 4 billion years or something. So something like `assertLt(governorBravo.votingDelay(), MAXIMUM_REASONABLE_TIME_PERIOD)

File: test/PooltogetherGovernor.t.sol [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import {PooltogetherGovernorTest} from "test/helpers/PooltogetherGovernorTest.sol";
+
+contract Constructor is PooltogetherGovernorTest {
+  function testFuzz_CorrectlySetsAllConstructorArgs(uint256 _blockNumber) public {
+    assertEq(governorBravo.name(), "Pooltogether Governor Bravo");
+    assertEq(address(governorBravo.token()), POOL_TOKEN);
+    // These values were all copied directly from the mainnet alpha governor at:
+    // 0xB3a87172F555ae2a2AB79Be60B336D2F7D0187f0
+    assertEq(INITIAL_VOTING_DELAY, 1);
```
Comment (apbendi) [inline test/PooltogetherGovernor.t.sol]: We have this asserts in the GC codebase but seeing them here I don't see what good they do. I think we can remove them for the other checks described in other comments below.

**Review by apbendi — COMMENTED** [review]
Mostly small stuff, overall looks good!

File: src/interfaces/IPOOL.sol (hunk 1 above)
Comment (alexkeating) [reply src/interfaces/IPOOL.sol:4]: Oh I didn't know that was a thing. I verified visually

Comment (alexkeating) [reply src/interfaces/IPOOL.sol:4]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/9/commits/22bfdac495003fc5a43b86e3c0d63226828a25c6

File: test/PooltogetherGovernor.t.sol (hunk 5 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/9/commits/2f07ad167101bbab14f62020236c56e22f6c1b3c

File: test/PooltogetherGovernor.t.sol (hunk 4 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol:14]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/9/commits/2f07ad167101bbab14f62020236c56e22f6c1b3c

File: src/interfaces/IPOOL.sol (hunk 1 above)
Comment (apbendi) [reply src/interfaces/IPOOL.sol:4]: It's a great feature!

File: test/PooltogetherGovernor.t.sol (hunk 4 above)
Comment (apbendi) [reply test/PooltogetherGovernor.t.sol:14]: Hmm, I see the constant but I don't see the assertions. Am I missing something @alexkeating?

File: test/helpers/PooltogetherGovernorTest.sol [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+import {Deploy} from "script/Deploy.s.sol";
+import {DeployInput} from "script/DeployInput.sol";
+import {IPOOL} from "src/interfaces/IPOOL.sol";
+import {PooltogetherGovernor} from "src/PooltogetherGovernor.sol";
+
+abstract contract PooltogetherGovernorTest is Test, DeployInput {
+  using FixedPointMathLib for uint256;
+
+  uint256 constant QUORUM = 2_500_000e18;
```
Comment (apbendi) [inline test/helpers/PooltogetherGovernorTest.sol]: The value for quorum which I see in the live GovernorAlpha is `100000000000000000000000` i.e. `100000e18`. We should use that value for now and check with PT that they want to keep it the same.

Comment (apbendi) [reply test/helpers/PooltogetherGovernorTest.sol]: Did catch this in the previous PR but it looks like our value for the proposal threshold is also incorrect. The proposal threshold I'm seeing (double check me) is `10000e18` while we have `1000e18`

File: test/PooltogetherGovernor.t.sol (hunk 4 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol:14]: You see correct, I am not sure what happened, but will get the changes up

File: test/helpers/PooltogetherGovernorTest.sol (hunk 6 above)
Comment (alexkeating) [reply test/helpers/PooltogetherGovernorTest.sol]: Double checked and your number is right

Comment (alexkeating) [reply test/helpers/PooltogetherGovernorTest.sol]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/9/commits/4a590b49a0a8ba2137e64890da3fe0745c379d57

File: test/PooltogetherGovernor.t.sol (hunk 4 above)
Comment (alexkeating) [reply test/PooltogetherGovernor.t.sol:14]: https://github.com/ScopeLift/pooltogether-governor-upgrade/pull/9/commits/4a590b49a0a8ba2137e64890da3fe0745c379d57

**Review by apbendi — APPROVED** [review]
This is good to go after we resolve the CI configuration issues. Thanks @alexkeating

**Pull request merged**